File: ios/CreateWallet.swift
//
//  CreateWallet.swift
//  WalletNativeDemo
//
//  Created by user on 02/03/23.
//

import UIKit
import HDWalletKit
import CryptoSwift
import React
import HDWalletKitBTC
import WalletCore

@objc(CreateWallet)
class CreateWallet: NSObject {

  @objc(generateMnemonics:)
 func generateMnemonics(_ callback: @escaping RCTResponseSenderBlock) {
   // Date is ready to use!
   let mnemonics = Mnemonic.create()
   let seed = Mnemonic.createSeed(mnemonic:mnemonics)
   let ethObj = generateAddress(seed: seed, coin: .ethereum)
   let btcObj = generateBip84Address(seed: seed, network: .main)
   let ltcObj = generateBip84Address(seed: seed, network: .ltc)
   let bchObj = generateBip84Address(seed: seed, network: .bch)
//   let dogeObj = generateDogeAddress(seed: seed, network: .doge, coin: .dogecoin)


   var obj = [String: Any]()
   obj["mnemonics"] =  mnemonics
   obj["eth"] = ethObj
   obj["btc"] = btcObj
   obj["ltc"] = ltcObj
   obj["bch"] = bchObj
//   obj["doge"] = dogeObj
   callback([String(data: try! JSONSerialization.data(withJSONObject: obj, options: .prettyPrinted), encoding: .utf8)!])
 }
  
  @objc(generateAddressFromMnemonics:callback:)
  func generateAddressFromMnemonics(menmonicsString:String, callback: @escaping RCTResponseSenderBlock) {
    
    let seed = Mnemonic.createSeed(mnemonic:menmonicsString)
    
    let ethObj = generateAddress(seed: seed, coin: .ethereum)
    let btcObj = generateBip84Address(seed: seed, network: .main)
    let ltcObj = generateBip84Address(seed: seed, network: .ltc)
    let bchObj = generateBip84Address(seed: seed, network: .bch)

//    let dogeObj = generateDogeAddress(seed: seed, network: .doge, coin: .dogecoin)

    var obj = [String: [String: String]]()
    obj["eth"] = ethObj
    obj["btc"] = btcObj
    obj["ltc"] = ltcObj
    obj["bch"] = bchObj
//    obj["doge"] = dogeObj
    callback([String(data: try! JSONSerialization.data(withJSONObject: obj, options: .prettyPrinted), encoding: .utf8)!])
  }
  
  func generateBip84Address(seed: Data, network: Network)-> [String: String]{
    let wallet = WalletBTC(seed: seed, network: network)
    let masterPrivateKey = wallet.privateKey
    let purpose = masterPrivateKey.derived(at: 84, hardens: true)
    let coinType = purpose.derived(at: network.coinType, hardens: true)
    let account = coinType.derived(at: 0, hardens: true)
    let change = account.derived(at: 0)
    let firstPrivateKey = change.derived(at: 0)
    let pvtKey = firstPrivateKey.wifiCompressed()
    let publicAddress = wallet.generateAddressBIP84(at: 0)
    
    var obj = [String: String]()
    obj["address"] = publicAddress
    obj["pvtKey"] = pvtKey
    return obj
  }
  
//  func generateDogeAddress(seed: Data, network: Network, coin: Coin)-> [String: String]{
//    let wallet = Wallet(seed: seed, network: network)
//    let masterPrivateKey = wallet.privateKey
//    let purpose = masterPrivateKey.derived(at: 44, hardens: true)
//    let coinType = purpose.derived(at: network.coinType, hardens: true)
//    let account = coinType.derived(at: 0, hardens: true)
//    let change = account.derived(at: 0)
//    let firstPrivateKey = change.derived(at: 0)
//    let pvtKey = firstPrivateKey.wifiCompressed()
//
//    let walletNew = Wallet(seed: seed, coin: coin)
//    let accountNew = walletNew.generateAccount()
//    let publicAddress = accountNew.address
//
//    var obj = [String: String]()
//    obj["address"] = publicAddress
//    obj["pvtKey"] = pvtKey
//    return obj
//  }
  
 
 
  func generateAddress(seed: Data, coin: Coin) -> [String: String]{
    let wallet = Wallet(seed: seed, coin: coin)
    let account = wallet.generateAccount()
    let publicAddress = account.address
    let pvtKey = account.rawPrivateKey
    var obj = [String: String]()
    obj["address"] = publicAddress
    obj["pvtKey"] = coin == .ethereum ? "0x"+pvtKey : pvtKey
    return obj
  }
  
  
  
   @objc(signSolanaTransaction:toAddress:amount:recentBlockhash:callback:)
   func signSolanaTransaction(
     pvtKey: String,
     toAddress: String,
     amount: String, // Amount in lamports (1 SOL = 1,000,000,000 lamports)
     recentBlockhash: String,
     _ callback: @escaping RCTResponseSenderBlock
   ) {
     
 //    let privateKey = PrivateKey(data: Base58.decode(pvtKey)!)
     let hdWallet = HDWallet(mnemonic: pvtKey, passphrase: "")!
     let privateKey = hdWallet.getKeyForCoin(coin: .solana)
         
     let transferMessage = SolanaTransfer.with {
       $0.recipient = toAddress
       $0.value = UInt64(amount)!
     }
     
     let input = SolanaSigningInput.with {
                 $0.transferTransaction = transferMessage
                 $0.recentBlockhash = recentBlockhash
                 $0.privateKey = privateKey.data
                 $0.txEncoding = .base64
               }
     let output: SolanaSigningOutput = AnySigner.sign(input: input, coin: .solana)
      
     print("Signed Transaction (Hex): \(output.encoded)")
     
     let obj = ["rawTxn": output.encoded, "signature": output.signatures.first?.signature ]
     
     print(obj)
     callback([String(data: try! JSONSerialization.data(withJSONObject: obj, options: .prettyPrinted), encoding: .utf8)!])

   }
   
   @objc(signSolanaTokenTransaction:fromAddress:toAddress:tokenMintAddress:amount:recentBlockhash:decimal:callback:)
   func signSolanaTokenTransaction(pvtKey: String,
                                   fromAddress: String,
                                   toAddress: String,
                                   tokenMintAddress: String,
                                   amount: String, // Amount in lamports (1 SOL = 1,000,000,000 lamports)
                                   recentBlockhash: String,
                                   decimal: String,
                                   _ callback: @escaping RCTResponseSenderBlock) {
     let hdWallet = HDWallet(mnemonic: pvtKey, passphrase: "")!
     let privateKey = hdWallet.getKeyForCoin(coin: .solana)
     
     
     let fromTokenAccount = fromAddress//getTokenAccount(tokenAddress: tokenMintAddress, walletAddress: fromAddress)
     
     let toTokenAccount = toAddress//getTokenAccount(tokenAddress: tokenMintAddress, walletAddress: toAddress)
     
     let tokenTransferMessage = SolanaTokenTransfer.with {
       $0.tokenMintAddress = tokenMintAddress
       $0.recipientTokenAddress = toTokenAccount
       $0.senderTokenAddress = fromTokenAccount
       $0.amount = UInt64(amount)!
       $0.decimals = UInt32(decimal)!
     }
     let input = SolanaSigningInput.with {
       $0.tokenTransferTransaction = tokenTransferMessage
       $0.recentBlockhash = recentBlockhash
       $0.privateKey = privateKey.data
       $0.txEncoding = .base64
     }
     
     let output: SolanaSigningOutput = AnySigner.sign(input: input, coin: .solana)
     
     let obj = ["rawTxn": output.encoded, "signature": output.signatures.first?.signature ]
     
     print(obj)
     callback([String(data: try! JSONSerialization.data(withJSONObject: obj, options: .prettyPrinted), encoding: .utf8)!])
   }
   
   @objc(createTokenAccountSigner:mainAddress:tokenMintAddress:tokenAddress:recentBlockhash:callback:)
   func createTokenAccountSigner(pvtKey: String, mainAddress: String, tokenMintAddress: String, tokenAddress: String, recentBlockhash: String, _ callback: @escaping RCTResponseSenderBlock) {
     let hdWallet = HDWallet(mnemonic: pvtKey, passphrase: "")!
     let privateKey = hdWallet.getKeyForCoin(coin: .solana)
     let createAccountMessage = SolanaCreateTokenAccount.with {
               $0.mainAddress = mainAddress
               $0.tokenMintAddress = tokenMintAddress
               $0.tokenAddress = tokenAddress
           }
           
     
     let input = SolanaSigningInput.with {
       $0.createTokenAccountTransaction = createAccountMessage
       $0.recentBlockhash = recentBlockhash
       $0.privateKey = privateKey.data
       $0.txEncoding = .base64
       $0.priorityFeePrice = SolanaPriorityFeePrice.with {
         $0.price = 10000
       }
       $0.priorityFeeLimit = SolanaPriorityFeeLimit.with{
         $0.limit = 1000000
       }
     }

     let output: SolanaSigningOutput = AnySigner.sign(input: input, coin: .solana)
     let obj = ["rawTxn": output.encoded, "signature": output.signatures.first?.signature ]
     
   
     print(obj)
     callback([String(data: try! JSONSerialization.data(withJSONObject: obj, options: .prettyPrinted), encoding: .utf8)!])
   }
   
   @objc(signSolanaWalletConnectRequest:requestPayload:callback:)
  func signSolanaWalletConnectRequest(pvtKey: String, requestPayload: String, _ callback: @escaping RCTResponseSenderBlock)  {
    let hdWallet = HDWallet(mnemonic: pvtKey, passphrase: "")!
    let privateKey = hdWallet.getKeyForCoin(coin: .solana)
    
    let payload = String(format: """
         {"transaction":"%@"}
         """, requestPayload)
    // Step 1: Parse a signing request received through WalletConnect.
    let parsingInput = WalletConnectParseRequestInput.with {
      $0.method = .solanaSignTransaction
      $0.payload = payload
    }
    
    do {
      let parsingInputBytes = try parsingInput.serializedData()
      
      let parsingOutputBytes = WalletConnectRequest.parse(coin: .solana, input: parsingInputBytes)
      let parsingOutput = try WalletConnectParseRequestOutput(serializedBytes: parsingOutputBytes)
      
      var signingInput = parsingOutput.solana
      
      // Step 2: Set missing fields.
      signingInput.privateKey = privateKey.data
      signingInput.txEncoding = .base64
      
      // Step 3: Sign the transaction.
      let output: SolanaSigningOutput = AnySigner.sign(input: signingInput, coin: .solana)
      let obj = ["rawTxn": output.encoded, "signature": output.signatures.first?.signature ]
      
      print(obj)
      callback([String(data: try! JSONSerialization.data(withJSONObject: obj, options: .prettyPrinted), encoding: .utf8)!])
    } catch {
      //handle error
      print(error)
    }
    
    
    
    
  }

}

struct SolanaTransaction {
    struct TransferInstruction {
        let from: Data
        let to: Data
        let value: UInt64

        func serialize() -> Data {
            // Construct the byte array for the transfer instruction here.
            // This involves Solana-specific byte encoding.
            // You will need to follow the Solana Transfer Instruction format.
            return Data() // Placeholder
        }
    }
    
    struct Message {
        let recentBlockhash: String
        let instructions: [TransferInstruction]

        func serialize() -> Data {
            // Serialize the message into the proper format for Solana.
            return Data() // Placeholder
        }
    }
    
    var message: Message
    var signatures: [Data] = []

    init(message: Message) {
        self.message = message
    }

    mutating func addSignature(signature: Data) {
        signatures.append(signature)
    }

    func serialize() -> Data {
        // Serialize the entire transaction, including signatures.
        return Data() // Placeholder
    }
}

File: android/app/src/main/java/com/wallet/qubetics/CreateWallet.java
package com.wallet.qubetics;
import static android.content.ContentValues.TAG;

import android.os.Environment;
import android.util.Log;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import org.json.JSONException;
import org.json.JSONObject;


import java.io.File;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;

import com.google.protobuf.ByteString;

import wallet.core.java.AnySigner;
import wallet.core.jni.Base58;
import wallet.core.jni.CoinType;
import wallet.core.jni.HDWallet;
import wallet.core.jni.PrivateKey;
import wallet.core.jni.SolanaAddress;
import wallet.core.jni.WalletConnectRequest;
import wallet.core.jni.proto.Common;
import wallet.core.jni.proto.Solana;
import wallet.core.jni.proto.WalletConnect;

public class CreateWallet extends ReactContextBaseJavaModule {
    CreateWallet(ReactApplicationContext context) {
        super(context);
        System.loadLibrary("TrustWalletCore");
    }

    @NonNull
    @Override
    public String getName() {
        return "CreateWallet";
    }

    @ReactMethod
    public void generateMnemonics(Callback callback)  {
        Log.d("CreateWallet", "Create event called with name: ");
        HDWallet wallet = new HDWallet(128,"");

        String mnemonics = wallet.mnemonic();
        JSONObject addresses = generateAdresses(mnemonics);
        callback.invoke(addresses.toString());

    }

    @ReactMethod
    public void generateAddressFromMnemonics(String mnemonics, Callback callback) {
        JSONObject addresses = generateAdresses(mnemonics);
        Log.d("CreateWallet", addresses.toString());
        callback.invoke(addresses.toString());
    }

    @ReactMethod
    public void clearGarbageCollection() {
        Log.d("clearGarbageCollection", "clearGarbageCollection called");
        System.runFinalization();
        Runtime.getRuntime().gc();
        System.gc();
    }

    @ReactMethod
    public void signSolanaTransaction(String pvtKey, String toAddress, String amount, String recentBlockhash, Callback callback){
        HDWallet hdWallet = new HDWallet(pvtKey, "");
        CoinType coinSol = CoinType.SOLANA;
        PrivateKey solanaPrivateKey = hdWallet.getKeyForCoin(coinSol);
        long amountLong = Long.parseLong(amount);
        Solana.Transfer transferMessage = Solana.Transfer.newBuilder()
                .setRecipient(toAddress)
                .setValue(amountLong)
                .build();
        // Create a signing input
        Solana.SigningInput signingInput = Solana.SigningInput.newBuilder()
                .setTransferTransaction(transferMessage)
                .setRecentBlockhash(recentBlockhash)
                .setPrivateKey(ByteString.copyFrom(solanaPrivateKey.data()))
                .setTxEncoding(Solana.Encoding.Base64)
                .build();
        try {
            // Sign the transaction
            Solana.SigningOutput output = AnySigner.sign(signingInput, CoinType.SOLANA, Solana.SigningOutput.parser());
            wallet.core.jni.proto.Solana.PubkeySignature signature = output.getSignatures(0);
            final JSONObject txnObj = new JSONObject();
            txnObj.put("rawTxn", output.getEncoded());
            txnObj.put("signature", signature.getSignature());
            callback.invoke(txnObj.toString());
        } catch (Exception e) {
            // Handle it.
            Log.e("Sign Coin Txn Error", e.toString());
        }
    }

    @ReactMethod
    public void signSolanaTokenTransaction(String pvtKey, String fromAddress, String toAddress, String tokenMintAddress, String amount, String recentBlockhash, String decimal, Callback callback) {
        HDWallet hdWallet = new HDWallet(pvtKey, "");
        CoinType coinSol = CoinType.SOLANA;
        PrivateKey solanaPrivateKey = hdWallet.getKeyForCoin(coinSol);
        long amountLong = Long.parseLong(amount);
        int decimalInt = Integer.parseInt(decimal);
//        String senderTokenAddress = getTokenAccount(tokenMintAddress, fromAddress);
//        String recipientTokenAddress = getTokenAccount(tokenMintAddress, toAddress);

        Solana.TokenTransfer tokenTransferMessage = Solana.TokenTransfer.newBuilder()
                .setTokenMintAddress(tokenMintAddress)
                .setSenderTokenAddress(fromAddress)
                .setRecipientTokenAddress(toAddress)
                .setAmount(amountLong) // Represents 0.004 since decimals = 6
                .setDecimals(decimalInt)
                .build();

        // Create a signing input for the token transfer
        Solana.SigningInput signingInput = Solana.SigningInput.newBuilder()
                .setTokenTransferTransaction(tokenTransferMessage)
                .setRecentBlockhash(recentBlockhash)
                .setPrivateKey(ByteString.copyFrom(solanaPrivateKey.data()))
                .setTxEncoding(Solana.Encoding.Base64)
                .build();

        try {
            // Sign the transaction
            Solana.SigningOutput output = AnySigner.sign(signingInput, CoinType.SOLANA, Solana.SigningOutput.parser());
            wallet.core.jni.proto.Solana.PubkeySignature signature = output.getSignatures(0);
            final JSONObject txnObj = new JSONObject();
            txnObj.put("rawTxn", output.getEncoded());
            txnObj.put("signature", signature.getSignature());
            callback.invoke(txnObj.toString());
        } catch (Exception e) {
            // Handle it.
            Log.e("Sign Token Txn Error", e.toString());
        }
    }

    @ReactMethod
    public void createTokenAccountSigner(String pvtKey, String mainAddress, String tokenMintAddress, String tokenAddress, String recentBlockhash, Callback callback) {
        HDWallet hdWallet = new HDWallet(pvtKey, "");
        CoinType coinSol = CoinType.SOLANA;
        PrivateKey solanaPrivateKey = hdWallet.getKeyForCoin(coinSol);
        Solana.CreateTokenAccount createAccountMessage = Solana.CreateTokenAccount.newBuilder()
                .setMainAddress(mainAddress)
                .setTokenMintAddress(tokenMintAddress)
                .setTokenAddress(tokenAddress)
                .build();

        // Create the signing input
        Solana.SigningInput signingInput = Solana.SigningInput.newBuilder()
                .setCreateTokenAccountTransaction(createAccountMessage)
                .setRecentBlockhash(recentBlockhash)
                .setPrivateKey(ByteString.copyFrom(solanaPrivateKey.data()))
                .setTxEncoding(Solana.Encoding.Base64)
                .setPriorityFeePrice(Solana.PriorityFeePrice.newBuilder().setPrice(10000).build())
                .setPriorityFeeLimit(Solana.PriorityFeeLimit.newBuilder().setLimit(1000000).build())
                .build();

        try {
            // Sign the transaction
            Solana.SigningOutput output = AnySigner.sign(signingInput, CoinType.SOLANA, Solana.SigningOutput.parser());
            wallet.core.jni.proto.Solana.PubkeySignature signature = output.getSignatures(0);
            final JSONObject txnObj = new JSONObject();
            txnObj.put("rawTxn", output.getEncoded());
            txnObj.put("signature", signature.getSignature());
            callback.invoke(txnObj.toString());
        } catch (Exception e) {
            // Handle it.
            Log.e("Sign Token Txn Error", e.toString());
        }
    }
    @ReactMethod
    public void signSolanaWalletConnectRequest(String pvtKey, String requestPayload, Callback callback){
        HDWallet hdWallet = new HDWallet(pvtKey, "");
        CoinType coinSol = CoinType.SOLANA;
        PrivateKey solanaPrivateKey = hdWallet.getKeyForCoin(coinSol);

        String payload = "{\"transaction\":\"\"}";
        String finalPayload = payload.substring(0, 16) + requestPayload + payload.substring(16);
        System.out.println("Signed transaction finalPayload: " + finalPayload);
        // Step 1: Parse a signing request received through WalletConnect.
        WalletConnect.ParseRequestInput parsingInput = WalletConnect.ParseRequestInput.newBuilder()
                .setMethod(WalletConnect.Method.SolanaSignTransaction)
                .setPayload(finalPayload)
                .build();

        // Step 2: Parse the WalletConnect request.
        byte[] parsingOutputBytes = WalletConnectRequest.parse(CoinType.SOLANA, parsingInput.toByteArray());
        WalletConnect.ParseRequestOutput parsingOutput;
        try {
            parsingOutput = WalletConnect.ParseRequestOutput.parseFrom(parsingOutputBytes);
        } catch (Exception e) {
            throw new RuntimeException("Failed to parse output: " + e.getMessage(), e);
        }

        // Check for parsing errors.
        if (parsingOutput.getError() != Common.SigningError.OK) {
            throw new RuntimeException("Parsing error: " + parsingOutput.getError());
        }

        // Step 3: Set missing fields for the signing input.
        Solana.SigningInput signingInput = parsingOutput.getSolana().toBuilder()
                .setPrivateKey(ByteString.copyFrom(solanaPrivateKey.data()))
                .setTxEncoding(Solana.Encoding.Base64)
                .build();

        try {
            // Step 4: Sign the transaction.
            Solana.SigningOutput output = AnySigner.sign(signingInput, CoinType.SOLANA, Solana.SigningOutput.parser());
            wallet.core.jni.proto.Solana.PubkeySignature signature = output.getSignatures(0);
            final JSONObject txnObj = new JSONObject();
            txnObj.put("rawTxn", output.getEncoded());
            txnObj.put("signature", signature.getSignature());
            callback.invoke(txnObj.toString());
        }catch (Exception e) {
            Log.e("Sign Wallet Connect Txn Error", e.toString());
        }


    }

    public JSONObject generateAdresses(String mnemonics) {
        try {
            HDWallet hdWallet = new HDWallet(mnemonics, "");
            CoinType coinEth = CoinType.ETHEREUM;
            CoinType coinBtc = CoinType.BITCOIN;

            String ethSeeds = null;
            String btcSeeds = null;

            byte[] solanaPrivateKey = getSolanaPrivateKey(hdWallet);

            // Convert to a hex string for easy viewing or storage
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
                ethSeeds = Base64.getEncoder().encodeToString(hdWallet.getKeyForCoin(coinEth).data());
                btcSeeds = toWIF(hdWallet.getKeyForCoin(coinBtc).data(), true);
            }

            final JSONObject objectBTC = new JSONObject();
            final JSONObject objectETH = new JSONObject();
            try {
                objectBTC.put("address", hdWallet.getAddressForCoin(coinBtc));
                objectBTC.put("pvtKey", btcSeeds);
            } catch (JSONException e) {
                Log.e(TAG, "Failed to create JSONObject", e);
            }
            try {
                objectETH.put("address", hdWallet.getAddressForCoin(coinEth));
                objectETH.put("pvtKey", ethSeeds);
            } catch (JSONException e) {
                Log.e(TAG, "Failed to create JSONObject", e);
            }



            final JSONObject addresses = new JSONObject();
            try {
                // With put you can add a name/value pair to the JSONObject
                addresses.put("eth", objectETH);
                addresses.put("btc", objectBTC);
                addresses.put("mnemonics", mnemonics);
            } catch (JSONException e) {
                Log.e(TAG, "Failed to create JSONObject", e);
            }

            return addresses;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static String toWIF(byte[] keyData, boolean compressed) {
        // Add version prefix (0x80 for Bitcoin mainnet)
        byte[] versionedKey = new byte[keyData.length + 1];
        versionedKey[0] = (byte) 0x80;
        System.arraycopy(keyData, 0, versionedKey, 1, keyData.length);

        // Add compression flag (0x01) if compressed
        byte[] keyWithCompression;
        if (compressed) {
            keyWithCompression = new byte[versionedKey.length + 1];
            System.arraycopy(versionedKey, 0, keyWithCompression, 0, versionedKey.length);
            keyWithCompression[versionedKey.length] = 0x01;
        } else {
            keyWithCompression = versionedKey;
        }

        // Perform double SHA-256 for checksum
        MessageDigest sha256Digest = null;//getInstance("SHA-256");
        try {
            sha256Digest = MessageDigest.getInstance("SHA-256");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
        byte[] hash1 = sha256Digest.digest(keyWithCompression);
        byte[] hash2 = sha256Digest.digest(hash1);

        // Take the first 4 bytes as the checksum
        byte[] checksum = new byte[4];
        System.arraycopy(hash2, 0, checksum, 0, 4);

        // Concatenate key with checksum
        byte[] wifBytes = new byte[keyWithCompression.length + checksum.length];
        System.arraycopy(keyWithCompression, 0, wifBytes, 0, keyWithCompression.length);
        System.arraycopy(checksum, 0, wifBytes, keyWithCompression.length, checksum.length);

        // Encode the result to Base58
        return Base58.encodeNoCheck(wifBytes);
    }

    public static byte[] getSolanaPrivateKey(HDWallet hdWallet) {
        // Get the private key for Solana from the HD wallet
        PrivateKey solanaPrivateKey = hdWallet.getKeyForCoin(CoinType.SOLANA);

        // Extract the raw private key bytes (32 bytes for the private key)
        byte[] privateKeyData = solanaPrivateKey.data();

        // Check the length of the private key data
        if (privateKeyData.length != 32) {
            throw new IllegalArgumentException("Invalid Solana private key length: " + privateKeyData.length);
        }

        // For Solana, you usually represent the private key with the concatenation of the private and public keys.
        // The public key can be derived using the private key.
        byte[] publicKey = solanaPrivateKey.getPublicKeyEd25519().data();

        // Combine the private key and public key into a 64-byte array
        byte[] solanaKeyPair = new byte[64];
        System.arraycopy(privateKeyData, 0, solanaKeyPair, 0, 32); // Copy private key
        System.arraycopy(publicKey, 0, solanaKeyPair, 32, 32); // Copy public key

        return solanaKeyPair; // Return the combined 64-byte key pair
    }

    public static String solanaPrivateKeyToHex(byte[] solanaKeyPair) {
//        StringBuilder hexString = new StringBuilder();
//        for (byte b : solanaKeyPair) {
//            hexString.append(String.format("%02x", b));
//        }
        return Base58.encodeNoCheck(solanaKeyPair);
    }

    public void writeFile(String PATH) {
        File directory = new File(PATH);
        if (!directory.exists()) {
            directory.mkdir();
            // If you require it to make the entire directory path including parents,
            // use directory.mkdirs(); here instead.
        }
    }

    static String getTokenAccount(String tokenAddress, String walletAddress) {
        SolanaAddress address = new SolanaAddress(walletAddress);
        return address.defaultTokenAddress(tokenAddress);
    }
}
